schema/pgsql/schema.sql

```diff
@@ -258,25 +316,35 @@ CREATE TABLE rule (
     name citext NOT NULL,
     timeperiod_id bigint REFERENCES timeperiod(id),
     object_filter text,
-    is_active boolenum NOT NULL DEFAULT 'y',
+
+    changed_at bigint NOT NULL DEFAULT EXTRACT(EPOCH FROM NOW()) * 1000,
+    deleted boolenum NOT NULL DEFAULT 'n',
 
     CONSTRAINT pk_rule PRIMARY KEY (id)
 );
 
+CREATE INDEX idx_rule_changed_at ON rule(changed_at);
+
 CREATE TABLE rule_escalation (
     id bigserial,
     rule_id bigint NOT NULL REFERENCES rule(id),
-    position integer NOT NULL,
+    position integer,
     condition text,
     name citext, -- if not set, recipients are used as a fallback for display purposes
     fallback_for bigint REFERENCES rule_escalation(id),
 
+    changed_at bigint NOT NULL DEFAULT EXTRACT(EPOCH FROM NOW()) * 1000,
+    deleted boolenum NOT NULL DEFAULT 'n',
+
     CONSTRAINT pk_rule_escalation PRIMARY KEY (id),
 
-    UNIQUE (rule_id, position),
-    CHECK (NOT (condition IS NOT NULL AND fallback_for IS NOT NULL))
+    UNIQUE (rule_id, position), -- column must be NULLed for deletion via "deleted = 'y'"
+    CONSTRAINT ck_rule_escalation_fallback_has_no_fallback CHECK (NOT (condition IS NOT NULL AND fallback_for IS NOT NULL)),
```

Comment:
Fallback has no fallback? One of these words should be condition :sweat_smile:

Reply:
No Thoughts Head Empty.